sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+      this.source = source;
+      this.checkpointMark = checkpointMark;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory = source.connectionFactory;
+      try {
+        this.connection = connectionFactory.createConnection();
+        this.connection.start();
+        this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        if (source.topic != null) {
+          this.consumer = this.session.createConsumer(this.session.createTopic(source.topic));
+          this.destination = source.topic;
+        } else {
+          this.consumer = this.session.createConsumer(this.session.createQueue(source.queue));
+          this.destination = source.queue;
+        }
+
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        TextMessage message = (TextMessage) this.consumer.receiveNoWait();
+
+        if (message == null) {
+          return false;
+        }
+
+        Map<String, Object> headers = new HashMap<>();
+        headers.put("JMSMessageID", message.getJMSMessageID());
+        headers.put("JMSTimestamp", message.getJMSTimestamp());
+        headers.put("JMSCorrelationID", message.getJMSCorrelationID());
+        headers.put("JMSReplyTo", message.getJMSReplyTo());
+        headers.put("JMSDestination", message.getJMSDestination());
+        headers.put("JMSDeliveryMode", message.getJMSDeliveryMode());
+        headers.put("JMSRedelivered", message.getJMSRedelivered());
+        headers.put("JMSType", message.getJMSType());
+        headers.put("JMSExpiration", message.getJMSExpiration());
+        headers.put("JMSPriority", message.getJMSPriority());
+
+        Map<String, Object> properties = new HashMap<>();
+        Enumeration propertyNames = message.getPropertyNames();
+        while (propertyNames.hasMoreElements()) {
+          String propertyName = (String) propertyNames.nextElement();
+          properties.put(propertyName, message.getObjectProperty(propertyName));
+        }
+
+        JmsRecord jmsRecord = new JmsRecord(headers, properties, message.getText());
+
+        currentRecord = jmsRecord;
+        currentTimestamp = new Instant(message.getJMSTimestamp());
+
+        return true;
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public JmsRecord getCurrent() throws NoSuchElementException {
+      return currentRecord;
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return currentTimestamp;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      return currentTimestamp;
+    }
+
+    @Override
+    public CheckpointMark getCheckpointMark() {
+      return new JmsCheckpointMark(destination,
+          currentRecord.getHeaders().get("JMSMessageID").toString());
```

Comment:
another point in favor of either extracting these as fields (currentRecord.headers.messageId) or string constants :)